umitag.py

```diff
@@ -70,58 +73,95 @@ def get_molecular_barcode(read1, read2, index1, index2, strategy='8B12X,,,'):
+    space = query_name.split(' ')
+    pts = space[0].split(':')
+    pts[2] = '{}_{}'.format(pts[2], molecular_id)
+    query_name = '{}\n'.format(':'.join(pts))
+    if len(space) > 1:
+        query_name = '{} {}'.format(query_name.replace('\n', ''), ' '.join(space[1:]))
+    return query_name
+
+
 def process_fq(read1_out, read2_out, read1, read2, index1, index2, pattern, start, stop):
     r1_umitagged_unsorted_file = '{}_{}.tmp'.format(read1_out, start)
     r2_umitagged_unsorted_file = '{}_{}.tmp'.format(read2_out, start)
 
     # Create UMI-tagged R1 and R2 FASTQs
     r1_umitagged = open(r1_umitagged_unsorted_file, 'w')
     r2_umitagged = open(r2_umitagged_unsorted_file, 'w')
-    for r1, r2, i1, i2 in itertools.izip(fq(read1, start, stop), fq(read2, start, stop), fq(index1, start, stop), fq(index2, start, stop)):
-        # Create molecular ID by concatenating molecular barcode and beginning of r1 read sequence
-        molecular_id, r1[1], r2[1], r1[3], r2[3] = get_molecular_barcode(r1, r2, i1, i2, pattern)
-        # Add molecular id to read headers
-        r1[0] = '%s %s\n' % (r1[0].rstrip(), molecular_id)
-        r2[0] = '%s %s\n' % (r2[0].rstrip(), molecular_id)
-        r1[0] = r1[0].replace(' ', ':')
-        r2[0] = r2[0].replace(' ', ':')
-        for line in r1:
-            r1_umitagged.write(line)
-        for line in r2:
-            r2_umitagged.write(line)
-    r1_umitagged.close()
-    r2_umitagged.close()
+    try:
+        for r1, r2, i1, i2 in itertools.izip(fq(read1, start, stop), fq(read2, start, stop), fq(index1, start, stop), fq(index2, start, stop)):
+            # Create molecular ID by concatenating molecular barcode and beginning of r1 read sequence
+            molecular_id, r1[1], r2[1], r1[3], r2[3] = get_molecular_barcode(r1, r2, i1, i2, pattern)
+            # Add molecular id to read headers
+            r1[0] = tag_query_name(r1[0], molecular_id)
+            r2[0] = tag_query_name(r2[0], molecular_id)
+            for line in r1:
+                r1_umitagged.write(line)
+            for line in r2:
+                r2_umitagged.write(line)
+    except Exception as e:
+        print(e)
+        raise e
+    finally:
+        r1_umitagged.close()
+        r2_umitagged.close()
     return r1_umitagged_unsorted_file, r2_umitagged_unsorted_file
 
+
 def get_numlines(fpath):
     ct = 0
     with open(fpath, 'rb') as fh:
         for _ in fh:
             ct += 1
     return ct
 
+
+def concat(partial_file, aggregate_file):
+    for rfile in [partial_file, aggregate_file]:
+        if not os.path.exists(rfile):
+            raise ValueError('%s does not exist', rfile)
+    with open(aggregate_file, 'a') as tofile:
+        with open(partial_file, 'rb') as pfile:
+            tofile.writelines(pfile.readlines())
+
+
+def sort_fastqs(r_umitagged_unsorted_file, sort_opts, read_out):
+    cmd = 'cat {} | paste - - - - | sort -k1,1 {} | tr "\t" "\n" > {}'.format(r_umitagged_unsorted_file, sort_opts,
```

Comment:
Can you add a comment why you're calling out to sort, instead of sorting in python? Are inputs expected to be too large to fit in memory?

Reply:
It's just standard to use this command in the pipeline and I didn't want to change the original code, but it would be interesting to compare the efficiency.